packages/adapters-library/src/core/apy-calculators/BalanceOfApyCalculator.ts

```diff
@@ -0,0 +1,122 @@
+import { AVERAGE_BLOCKS_PER_DAY } from '../constants/AVERAGE_BLOCKS_PER_DAY'
+import { ApyCalculation, ApyCalculator, EvmApyArgs } from './ApyCalculator'
+import { computeApr, computeApy } from './common'
+
+/**
+ * Calculates the APY of a position by calculating the fees as the difference of balance in
+ * the underlying token between a given day and the day before.
+ *
+ * Is compatible with ANY DeFi protocol where the position is tracked by a single amount in wei (for instance the balanceOf an ERC20),
+ * and where the balance varies only due to accruing fees/interest ( assuming no deposit or withdrawal).
+ *
+ * We can translate this mathematically to:
+ * balance(t') = balance(t) + fees(t)
+ *
+ * Compatible protocols / products are:
+ * - Stablecoins
+ * - Lending (aToken, cToken, ...)
+ * - LSTs
+ */
+export class BalanceOfApyCalculator implements ApyCalculator<EvmApyArgs> {
+  // Duration in days of the period where we look at the fees earned. The APY and APR are then annualized based on this period.
+  public static readonly PERIOD_DURATION_DAYS = 1
+
+  // How many periods there is in a year
+  public static readonly FREQUENCY = 365
+
+  public async getApy({
+    userAddress,
+    blockNumber,
+    protocolTokenAddress,
+    chainId,
+    adapter,
+  }: EvmApyArgs): Promise<ApyCalculation> {
+    const blocknumberEnd = blockNumber
+    const blocknumberStart =
+      blocknumberEnd -
+      AVERAGE_BLOCKS_PER_DAY[chainId] *
+        BalanceOfApyCalculator.PERIOD_DURATION_DAYS
+
+    const [positionsStart, positionsEnd] = await Promise.all([
+      adapter.getPositions({
+        userAddress,
+        blockNumber: blocknumberStart,
+        protocolTokenAddresses: [protocolTokenAddress],
+      }),
+      adapter.getPositions({
+        userAddress,
+        blockNumber: blocknumberEnd,
+        protocolTokenAddresses: [protocolTokenAddress],
+      }),
+    ])
+
+    const balanceStartWei =
+      positionsStart?.[0]?.tokens?.[0]?.tokens?.[0]?.balanceRaw
+    const balanceEndWei =
+      positionsEnd?.[0]?.tokens?.[0]?.tokens?.[0]?.balanceRaw
```

Comment:
If the underlying token length is more than 1 I think you could log/error/return _unsupported pool_

```
 if(positionsStart?.[0]?.tokens?.[0]?.tokens.length > 1) {exit...}
```

Reply:
This is an optimisation for another day but we should expose a way for you to be able to check if the protocol token has more than one underlying before doing the getPositions request

Reply:
e.g. we could add on here get pool metadata where you can check the number of underlying tokens

```
export type IMetadataProvider = {
  getMetadata: (input: {
    protocolId: Protocol
    productId: string
  }) => Promise<ProtocolToken[]>

  getPoolCount: (protocolId: Protocol, productId: string) => Promise<number>
}
```